framework/include/base/AuxGroupExecuteMooseObjectWarehouse.h

```diff
@@ -0,0 +1,118 @@
+
+/****************************************************************/
+/*               DO NOT MODIFY THIS HEADER                      */
+/* MOOSE - Multiphysics Object Oriented Simulation Environment  */
+/*                                                              */
+/*           (c) 2010 Battelle Energy Alliance, LLC             */
+/*                   ALL RIGHTS RESERVED                        */
+/*                                                              */
+/*          Prepared by Battelle Energy Alliance, LLC           */
+/*            Under Contract No. DE-AC07-05ID14517              */
+/*            With the U. S. Department of Energy               */
+/*                                                              */
+/*            See COPYRIGHT for full restrictions               */
+/****************************************************************/
+
+#ifndef AUXGROUPEXECUTEMOOSEOBJECTWAREHOUSEBASE_H
+#define AUXGROUPEXECUTEMOOSEOBJECTWAREHOUSEBASE_H
+
+// MOOSE includes
+#include "ExecuteMooseObjectWarehouse.h"
+#include "Postprocessor.h"
+
+class UserObject;
+
+/**
+ * General warehouse for storing MooseObjects based on relation to AuxKernel execution.
+ */
+template<typename T>
+class AuxGroupExecuteMooseObjectWarehouse : public ExecuteMooseObjectWarehouse<T>
+{
+
+public:
+
+  /// Using these from base class
+  using MooseObjectWarehouse<T>::checkThreadID;
+  using ExecuteMooseObjectWarehouse<T>::_all_objects;
+  using ExecuteMooseObjectWarehouse<T>::_execute_objects;
+  using ExecuteMooseObjectWarehouse<T>::_num_threads;
+
+  /**
+   * Constructor.
+   */
+  AuxGroupExecuteMooseObjectWarehouse(bool thread = true);
+
+  /**
+   * Access the AuxGroup via bracket operator.
+   */
+  const ExecuteMooseObjectWarehouse<T> & operator[](Moose::AuxGroup group) const;
+
+  /**
+   * Call this inorder to separate the stored objects into the various AuxGroup catagories.
+   *
+   * @see FEProblem::initialSetup()
+   */
+  void updateDependObjects(const std::set<std::string> & depend_uo, THREAD_ID tid = 0);
+
+  /**
+   * Performs a sort using the DependencyResolver.
+   */
+  void sort(THREAD_ID tid = 0);
+
+protected:
+
+  /// Storage for the group sorted objects
+  std::map<Moose::AuxGroup, ExecuteMooseObjectWarehouse<T> > _group_objects;
```

Comment:
You're using an enum as a key (0, 1, 2). You may as well use a vector, faster, less storage, etc.